### PR TITLE
[Feature] Soft deleting skills

### LINKED ISSUE #7576: [API Mutation] Skills can be soft deleted, without affecting past posters/profiles
## ✨ Feature

API mutation which lets Skills be soft-deleted by Platform Admins.

## 🕵️ Details

Platform Admins, and only platform admins, can delete existing skills. This fails (with a legible error) if the skill is required or optional for a currently-open pool.

Soft deleted skills should:
- continue to appear if on existing Job Posters
- continue to appear if on existing User Profiles
- continue to appear if on existing Applications

Soft deleted skills should NOT:
- be possible to add to new Job Posters
- be options when adding skills to skill library
- be options when adding skills to experiences in profile or on application
- be options on talent seeker filters
- be options on any other skill picker
- appear on the Skills table on admin side
- appear in any dropdowns on admin side (like for attaching to skill families)

## 🙋‍♀️ Proposed Implementation

Suggested: Use soft-deletes, the same as User.  Possibly a custom archived_at field (like Pools) if soft-deletes won't work well.

## ✅ Acceptance Criteria

- [ ] A mutation is added to delete Skills
  - [ ] this soft-deletes the skill
  - [ ] only Platform Admins have permission to do this
  - [ ] mutation fails with a legible error if the skill is used by any currently open pool
  - [ ] Publishing an existing draft pool, or re-opening any closed pool, fails with a legible error if it uses any soft-deleted skills

Soft deleted skills should:

- [ ] - continue to appear if on existing Job Posters
- [ ] - continue to appear if on existing User Profiles
- [ ] - continue to appear if on existing Applications

Soft deleted skills should NOT:

- [ ] - be possible to add to new Job Posters
- [ ] - be options when adding skills to skill library
- [ ] - be options when adding skills to experiences in profile
- [ ] - be options on talent seeker filters
- [ ] - be options on any other skill picker
- [ ] - appear on the Skills table on admin side
- [ ] - appear in any dropdowns on admin side (like for attaching to skill families)

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #7576

## 👋 Introduction

Introduces a mutation to soft delete skills. 

## 🕵️ Details

The mutation was the relatively straightforward part, going around testing and seeing changes or what is impacted was the trickier part. 
- Cannot delete a skill used by active posters
- Cannot publish a pool with deleted skills
- Cannot reopen a pool with deleted skills
- Things that query `skills` like a skill picker will not grab deleted skills
- `Userskill`, `Pool`, and `Experience` models can fetch their deleted skills

## 🧪 Testing

1. Exercise mutation below
2. Soft delete some skills and see if they are gone where they should be, but present where they still should be
3. Observe validation errors for certain things

Delete Skills with: 

```
mutation deleteSkill {
  deleteSkill(id: ":skillId") {
    id
  }
}
```

Fish for soft deleted skills with queries like: 

```
query me {
  me {
    userSkills {
      skill {
        id
      }
    }
  }
}
```

## 📸 Screenshot

image

### CODE REVIEW

File: api/graphql/schema.graphql [hunk 1]
```diff
@@ -1759,6 +1761,11 @@ type Mutation {
     @update
     @guard
     @can(ability: "update", find: "id")
+  deleteSkill(id: UUID! @whereKey): Skill
+    @delete
+    @validator(class: "App\\GraphQL\\Validators\\DeleteSkillValidator")
+    @can(ability: "delete", find: "id")
```
Comment (tristan-orourke) [inline api/graphql/schema.graphql:1767]: The SkillPolicy.delete method doesn't take a skill argument, so you can leave out `find` here.

File: api/tests/Feature/ExperienceTest.php [hunk 2]
```diff
@@ -474,4 +477,45 @@ public function testSkillRelationshipSkipsSoftDeletedPivots(): void
...
+        Skill::factory()->count(3)->create();
+        $experience = AwardExperience::factory()->withSkills(3)->create([
+            'user_id' => $this->platformAdmin->id,
+        ]);
+        $randomDeletedSkill = Skill::first();
+        DB::table('skills')
+            ->where('id', $randomDeletedSkill->id)
+            ->update([
+                'deleted_at' => config('constants.past_datetime'),
+            ]);
```
Comment (tristan-orourke) [inline api/tests/Feature/ExperienceTest.php:493]: This isn't wrong, but you could just use `->delete()`

File: api/tests/Feature/PoolTest.php [hunk 3]
```diff
@@ -551,4 +553,96 @@ public function testCantUnarchiveClosed(): void
...
+        // create complete but unpublished pool with a deleted skill
+        $classification = Classification::factory()->create();
+        $pool = Pool::factory()->published()->create([
+            'published_at' => null,
+            'closing_date' => config('constants.far_future_datetime'),
+        ]);
+        $pool->classifications()->sync([$classification->id]);
+        $skill1 = Skill::factory()->create();
+        $skill2 = Skill::factory()->create(['deleted_at' => config('constants.past_datetime')]);
+        $pool->essentialSkills()->sync([$skill1->id, $skill2->id]);
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/tests/Feature/PoolTest.php:568]: Oh cool, didn't think it actually worked like that.
I do like passing in ids to mirror `sync` for graphQL operations

File: api/tests/Feature/ExperienceTest.php (hunk 2 above)
Comment (vd1992) [reply api/tests/Feature/ExperienceTest.php]: Ah, I don't need the Facade at all here, I can use the other version of `delete()`
Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/a79de8ee608fd8973167be359ee6dfa661711581

File: api/app/GraphQL/Validators/DeleteSkillValidator.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            ->get()
+            ->load(['essentialSkills', 'nonessentialSkills']);
+
+        foreach ($activePools as $pool) {
+            $essentialSkillsIds = $pool->essentialSkills->pluck('id')->toArray();
+            $nonessentialSkillsIds = $pool->nonessentialSkills->pluck('id')->toArray();
+            $poolSkillsIds = array_merge($essentialSkillsIds, $nonessentialSkillsIds);
+
+            if (in_array($skillId, $poolSkillsIds)) {
+                throw ValidationException::withMessages(['SkillUsedByActivePoster']);
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/DeleteSkillValidator.php]: Extracted in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/e65d8826e28db21b36b73173e431983a984ec037

File: api/app/GraphQL/Mutations/ChangePoolClosingDate.php [hunk 5]
```diff
@@ -14,8 +17,32 @@ final class ChangePoolClosingDate
...
+        if ($pool->closing_date <= $now) {
+
+            $deletedSkillsIds = Skill::onlyTrashed()->get()->pluck('id')->toArray();
+            $essentialSkillsIds = $pool->essentialSkills->pluck('id')->toArray();
+            $nonessentialSkillsIds = $pool->nonessentialSkills->pluck('id')->toArray();
+            $poolSkillsIds = array_merge($essentialSkillsIds, $nonessentialSkillsIds);
+            $expiredPoolSkills = array_intersect($poolSkillsIds, $deletedSkillsIds);
+
+            if (count($expiredPoolSkills) > 0) {
+                throw ValidationException::withMessages(['CannotReopenUsingDeletedSkill']);
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/ChangePoolClosingDate.php]: Extracted in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/a2d78febc024dd54b22edbca30a3d6a838076546

File: api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php [hunk 6]
```diff
@@ -37,10 +39,22 @@ public function rules(): array
 
             // Essential skills and Asset skills
             'essential_skills' => ['required', 'array', 'min:1'],
-            'essential_skills.*.id' => ['required', 'uuid', 'exists:skills,id'],
+            'essential_skills.*.id' => Rule::forEach(function (?string $value, string $attribute) use ($deletedSkillsIds) {
+                return [
+                    'required',
+                    'uuid',
+                    'exists:skills,id',
+                    Rule::notIn($deletedSkillsIds),
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php]: Yeah, reusable indeed https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/609f083d49cf138dbb0d67eaeab1be6328b28ec7

File: api/app/Rules/SkillNotUsedByActivePool.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /**
+     * Run the validation rule.
+     */
+    public function validate(string $attribute, mixed $value, Closure $fail): void
+    {
+        // validation fails if skill is in use by an active pool
+        $skillId = $value;
+        $activePools = Pool::where('published_at', '<=', Carbon::now()->toDateTimeString())
+            ->where('closing_date', '>', Carbon::now()->toDateTimeString())
```
Comment (tristan-orourke) [inline api/app/Rules/SkillNotUsedByActivePool.php:21]: Feels like `currentlyActive` or `currentlyOpen` should be a scope

File: api/app/Rules/SkillNotUsedByActivePool.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            ->get()
+            ->load(['essentialSkills', 'nonessentialSkills']);
+
+        foreach ($activePools as $pool) {
```
Comment (tristan-orourke) [inline api/app/Rules/SkillNotUsedByActivePool.php:25]: I think you could do this faster by doing this as part of the query itself. Include a where clause which checks that the pool has `$skillId` in any of its essential skills or nonessential skills, and then just check if $activePools returned anything or not.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Nice work switching to rules and validators! Just a few suggestions now.

File: api/app/GraphQL/Validators/DeleteSkillValidator.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            'id' => [new SkillNotUsedByActivePool],
+        ];
+    }
+
+    /**
+     * Return the validation messages
+     */
+    public function messages(): array
+    {
+        return [];
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/DeleteSkillValidator.php:28]: I am letting the default message in the rule bubble up, defining in the messages array overrides it

image

image

And it should be working without issue

image

File: api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php [hunk 10]
```diff
@@ -62,6 +71,8 @@ public function messages(): array
             'advertisement_location.*.required_with' => 'You must enter both french and english fields for the advertisement_location',
             'in' => ':attribute does not contain a valid value.',
             'essential_skills.required' => 'EssentialSkillRequired',
+            'essential_skills.*.id.'.SkillNotDeleted::class => ApiErrorEnums::ESSENTIAL_SKILLS_CONTAINS_DELETED,
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php:74]: Laravel Pint won't allow it

File: api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php [hunk 11]
```diff
@@ -62,6 +71,8 @@ public function messages(): array
...
+            'nonessential_skills.*.id.'.SkillNotDeleted::class => ApiErrorEnums::NONESSENTIAL_SKILLS_CONTAINS_DELETED,
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php:75]: Laravel Pint won't allow it

File: api/app/GraphQL/Validators/DeleteSkillValidator.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/app/GraphQL/Validators/DeleteSkillValidator.php:28]: Ah, got it. Could you just leave a comment, something like "Leave array empty to use rules' default messages"

File: api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/app/GraphQL/Validators/Mutation/PublishPoolValidator.php:74]: Oh, this is the recommended formatting? Then nevermind, sorry 😄

File: api/app/Rules/SkillNotUsedByActivePool.php (hunk 7 above)
Comment (vd1992) [reply api/app/Rules/SkillNotUsedByActivePool.php]: Yeah, definitely
Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/39e1b082ceaea1013f475b787df0101484128abf and tested with https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/b46cb0fcf3ce168fcc108ea996892665f6d540ce

File: api/app/GraphQL/Validators/DeleteSkillValidator.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Validators/DeleteSkillValidator.php]: Aye aye
https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/c252b17de7e6afaecad6e8979e8dd96d57c01508

File: api/app/Rules/SkillNotUsedByActivePool.php (hunk 8 above)
Comment (vd1992) [reply api/app/Rules/SkillNotUsedByActivePool.php]: Yeah, I guess we can do that and condense the rule a bit. 
Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7762/commits/29d18f51a0eb4aa0e66deb2fc22644ea4cbae9fb

**Review by tristan-orourke — APPROVED** [review]
Great job!

**Pull request merged**